### PR TITLE
[REVIEW] Add CUDA Managed Memory (Unified Memory) Allocation Mode

### PR DESCRIPTION
**Pull request by harrism**
This PR adds an option to rmmInitialize to use `cudaMallocManaged` (+ `cudaMemPrefetchAsync`) as the base allocator for RMM rather than `cudaMalloc`.

```
rmmAllocationMode_t mode = CudaManagedMemory | PoolAllocation;
rmmOptions_t options {mode, 0, false};
rmmInitialize(&options);
```

The code above will initialize RMM with a managed memory pool. Removing `PoolAllocation` will cause it to allocate without a pool by calling directly to `cudaMallocManaged()`.

Tests are also expanded to run all tests with all allocation modes (gtests and pytests).

Note this depends on new functionality I added to cnmem with this PR (already merged): NVIDIA/cnmem#7

This PR is an improved version of a (never merged) cuDF PR rapidsai/cudf#422.

Support for multiple simultaneous pools with different modes is future work.

### CODE REVIEW

File: include/memory.h [hunk 1]
```diff
@@ -41,12 +41,25 @@ typedef enum
...
+ *
+ * These settings can be ORed together. For example to use a pool of managed
+ * memory, use `mode = PoolAllocation | CudaManagedMemory`.
+ * --------------------------------------------------------------------------**/
 typedef enum
 {
   CudaDefaultAllocation = 0,  //< Use cudaMalloc for allocation
-  PoolAllocation,             //< Use pool suballocation strategy
+  PoolAllocation        = 1,  //< Use pool suballocation strategy
+  CudaManagedMemory     = 2,  //< Use cudaMallocManaged rather than cudaMalloc
```
[the remark this answers is not in the document]
Comment (harrism) [reply include/memory.h]: What would it look like? Happy to do it as a separate PR if the API makes sense.

[the remark this answers is not in the document]
Comment (harrism) [reply include/memory.h]: Fixed.

File: include/memory.hpp [hunk 2]
```diff
@@ -126,13 +122,18 @@ inline rmmError_t alloc(T** ptr, size_t size, cudaStream_t stream, const char* f
 
   if (!ptr && !size) {
     return RMM_SUCCESS;
+  } else if( !size ) {
+    ptr[0] = NULL;
```
Comment (jrhemstad) [inline include/memory.hpp]:
```suggestion
    ptr[0] = nullptr;
```

File: include/memory.hpp [hunk 3]
```diff
@@ -167,15 +168,21 @@ inline rmmError_t realloc(T** ptr, size_t new_size, cudaStream_t stream,
...
+  if (rmm::Manager::usePoolAllocator()) {
     RMM_CHECK(rmm::Manager::getInstance().registerStream(stream));
     RMM_CHECK_CNMEM(cnmemFree(*reinterpret_cast<void**>(ptr), stream));
     RMM_CHECK_CNMEM(
         cnmemMalloc(reinterpret_cast<void**>(ptr), new_size, stream));
   } else {
     RMM_CHECK_CUDA(cudaFree(*ptr));
-    RMM_CHECK_CUDA(cudaMalloc(reinterpret_cast<void**>(ptr), new_size));
+    if (!new_size)
+      ptr[0] = NULL;
```
Comment (jrhemstad) [inline include/memory.hpp]:
```suggestion
      ptr[0] = nullptr;
```

**Pull request merged**